### DISCUSSION TITLE
Better understanding Client vs. Server

### DISCUSSION THREAD

**Discussion by pcar75** [jamulussoftware/jamulus#2530] [General]
As I started in "Mute On Entry" #2520  ,  I really wish to understand this better.

@dcorson-ticino-com    wrote 
> Muting is not a server function.

@pcar75   ?!?! How can Muting (channels, not Mute Myself) not be a server function. Though, it is set by a (client) user muting/panning his or other(s) channels, it is definitely the server that is adjusting the mix for that user. My understanding is therefore that the server integrates and resends not individual tracks but personnalized mix to each and every participants ... the server does the job. Which may explain the wish : for large ensembles and choirs to send one same standard mix to each user (thus lessening the power/performance requirement on the server).

@dcorson-ticino-com   wrote 

> The server does not know if someone is muted or not soloed or if the volume is just turned down.
> The server mixes, but has no idea of why the volume multiplier is 0.

@pcar75   (...)  In my mind , the client receives the mixed audio from the server; the client cannot unmix, recalibrate the volume/pan of each stems comprising the mix and remix the audio according to that user's channels settings ?!  Hence the server must do it and it generates as many mixes as there are users according to each own's channels settings (pan, volume, mute, solo, group) ! The server does not know 'why' a user's volume multiplier is X but it must take it into account to generate that version of a mix for that user !

What is wrong with that picture ?!?!  Otherwise, how does it really works ??

**Comment by pljones** [comment]
> How can Muting (channels, not Mute Myself) not be a server function.

The server only responds to client requests for mixing operations and those requests only include _level_.  So the only function the server has is to use the level requested by a client to adjust the mix.

If a client wants to mute itself, however, that isn't a server function - that's the client sending silence rather than a signal.

Every client receives a _different mix_ from the server.  So _every client's source sound_ must get to the server for other clients to be able to adjust the mix to their preference.  By sending silence, a client prevents other clients including it in their mix.

Hence "mute myself" can _only_ be a client function.

**Reply by pcar75** [reply]
As already stated,  

> (...) Muting (channels, not Mute Myself) ...

I have no questions regarding the workings of 'Mute Myself'.   I surmise that if the client is sending zeroes instead of one "no more audio contribution" message to the server, it even helps in insuring that the network uplink 'quality' is preserved and not wholly or partly reclassified to an inferior speed/bandwith/route  by different nodes along the internet hops.

> (...) those requests only include level.  (...)

Can you clarify what you mean by 'level' ,  If you mean only volume levels then how does 'Pan' work ?

**Reply by pljones** [reply]
>  If you mean only volume levels then how does 'Pan' work ?

Volume level only affects volume level.  Pan position is a separate value.

**Comment by pcar75** [comment]
Thanks !  Info much appreciated.
